src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py

```diff
@@ -0,0 +1,451 @@
+        "dimension of the array, but found a2(1,:) in "
+        "DOT_PRODUCT(a1(:,1), a2(1,:)).")
+    check_validate(code, expected, fortran_reader)
+
+
+def test_validate_array_full_slice(fortran_reader):
+    '''Test that the DotProduct2CodeTrans validate method produces the
+    expected exception when at least one of the arguments uses array
+    slice notation but the array slice is not for the full range of
+    the dimension.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(a1,a2)\n"
+        f"real,intent(in) :: a1(:,:), a2(:,:)\n"
+        f"real :: result\n"
+        f"result = dot_product(a1(2:4,1),a2(:,10))\n"
+        f"end subroutine\n")
+    expected = (
+        "The DotProduct2CodeTrans transformation only supports the "
+        "transformation of a dotproduct intrinsic with an argument not "
+        "an array slice if the argument is for the 1st dimension of "
+        "the array and is for the full range of that dimension, but "
+        "found a1(2:4,1) in DOT_PRODUCT(a1(2:4,1), a2(:,10)).")
+    check_validate(code, expected, fortran_reader)
+
+
+def test_validate_real(fortran_reader):
+    '''Test that the DotProduct2CodeTrans validate method raises the
+    expected exception if the datatype of the arguments is not
+    real.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(v1,v2)\n"
+        f"integer,intent(in) :: v1(:), v2(:)\n"
+        f"integer :: result\n"
+        f"result = dot_product(v1,v2)\n"
+        f"end subroutine\n")
+    expected = (
+        "The DotProduct2CodeTrans transformation only supports arrays of "
+        "real data, but found v1 of type INTEGER in DOT_PRODUCT(v1, v2).")
+    check_validate(code, expected, fortran_reader)
+
+
+def test_validate_get_array_bound(monkeypatch, fortran_reader):
+    '''Test that the DotProduct2CodeTrans validate method calls the
+    _get_array_bound method.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(v1,v2)\n"
+        f"real,intent(in) :: v1(:), v2(:)\n"
+        f"real :: result\n"
+        f"result = dot_product(v1,v2)\n"
+        f"end subroutine\n")
+    expected = "Transformation Error: dummy"
+
+    def dummy(_1, _2, _3):
+        '''Utility used to raise the required exception.'''
+        raise TransformationError("dummy")
+
+    monkeypatch.setattr(dotproduct2code_trans, "_get_array_bound", dummy)
+    check_validate(code, expected, fortran_reader)
+
+
+# DotProduct2CodeTrans class apply method
+
+def test_apply_calls_validate():
+    '''Test that the DotProduct2CodeTrans apply method calls the validate
+    method.
+
+    '''
+    trans = DotProduct2CodeTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.apply(None)
+    assert ("The supplied node argument is not a DOTPRODUCT operator, found "
+            "'NoneType'." in str(info.value))
+
+
+@pytest.mark.parametrize("dim1,dim2", [("10", "10"), (":", "10"), ("10", ":")])
```

Comment:
Could you extend this so that the bounds are given by References too?

Reply:
Extended.